100 vodafone model draft rapid response

edarin opened the pull request:
Scripts to compute the deterministic model and deploy it in the dashboard

edarin committed:
standardise plot format

edarin committed:
adapt terminology stock/flow

edarin committed:
propose first deterministic model

edarin committed:
draft model

edarin committed:
viz baselineflows_hat national totals

edarin committed:
Merge remote-tracking branch 'upstream/dev' into 100-kyivstar-model-draft-rapid-response

njirons committed:
code to estimate monthly flows

edarin committed:
remove custom geo

edarin committed:
estimate monthly flows induced pop estimates

edarin committed:
draft comp across methods

njirons committed:
hromada-level monthly flow estimates

added code for hromada-level monthly flow estimates (aggregating over age and sex); renamed old age-sex specific code file

edarin committed:
Merge branch '100-kyivstar-model-draft-rapid-response' of https://github.com/OxfordDemSci/UkraineNowPop into 100-kyivstar-model-draft-rapid-response

edarin committed:
old version of model with stocks and baseline

edarin committed:
focus on monthly flow prediction

edarin committed:
process eurostat and unhcr refugee age-sex profiles

edarin committed:
first round of population estimates

edarin committed:
get dtm data from their border crossing survey

edarin committed:
add calibration of inflows from abroad

edarin committed:
reorganise determinist model folder

edarin committed:
avoid recalibrating the inflows from abroad

edarin committed:
make config of nginx container cleaner

edarin committed:
adapt age labelling when last is not open

edarin committed:
comment out the 7 days window in migration query (was for facebook)

edarin committed:
prep oblast level data for the dasboard

edarin committed:
empty script for benchmark

edarin committed:
comment deterministic model

edarin committed:
remove for clarity

edarin committed:
prepare release standardisation

edarin committed:
prep becnhamrking script

alaparicioc committed:
Update 1_benchmark.R

Initial code for benchmar. It needs update based on meeting on 2025/07/29

alaparicioc commented on src/1_pop_data/60_refugee.R:
```diff
@@ -0,0 +1,254 @@
+eurostat_refugee <- eurostat_refugee |>
+  select(sex, age, geo, TIME_PERIOD, OBS_VALUE) |>
+  mutate(
+    t = as.Date(paste(TIME_PERIOD, "-01", sep = "")) + months(1),
```
months does not work for me. Besides, what is your intention here?: moving every Eurostat month forward by one month?

doug-leasure replied:
This line of code worked for me.

edarin replied:
Eurostat report refugees per month-period reported "MM-YYY". I've added "01" to match our way of reporting of population. But therefore postponed it by one month to cover the full month. Does it make sense?

alaparicioc replied:
Which package is the correct source of the 'months()' function? I suspect I may be calling the wrong one.

Also, about the one-month shift: Vodafone’s figures are reported on the first day of each month. Does that date represent the count for that same month, or is it another value that is been rolled forward?

alaparicioc commented on src/1_pop_data/60_refugee.R:
```diff
@@ -0,0 +1,254 @@
+  uncount(count, .id = "id_dup") |>
+  mutate(
+    pop = case_when(
+      a == 15 & id_dup == 1 ~ pop * 3 / 5,
```
This assumption could go to the README doc

alaparicioc reviewed (changes requested):
The function months did not work for me and I did not get what was the intention there. It is a small aspect but I could not continue assessing this

alaparicioc committed:
Update 1_benchmark.R

alaparicioc committed:
Update 1_benchmark.R

doug-leasure commented on src/2_model/mobile_phone/0_deterministic.R:
```diff
@@ -0,0 +1,998 @@
+# Prepare reference data by age and sex
+hromada_agesex <- hromada_pop |>
+  filter(hromada_code %in% unique(monthlyFlows$destination_hromada)) |>
+  left_join(agesex |>
```
Unintentional many-to-many left_join. Check that results are as expected and specify `relationship = "many-to-many"` if appropriate.

doug-leasure commented on src/2_model/mobile_phone/0_deterministic.R:
```diff
@@ -0,0 +1,998 @@
+    View()
+}
+
+# Step 4: Evaluate and visualise model -------------------------------
```
It would be much better to save these figures to jpeg so that we can review them and potentially put them into reports or presentations without needing to manually re-run the code every time.

edarin replied:
true. i tend to wait to see which one is the one we want to keep. for example ill store the one for the README.

doug-leasure replied:
No problem to just blast them all to disk and sort that out later.

doug-leasure commented on src/2_model/mobile_phone/1_benchmark.R:
```diff
@@ -0,0 +1,658 @@
+ dplyr::select(ADM1_PCODE, s_name, a_name, pop0)
+
+
+codps23 <- read.csv(file.path(data_dir, "COD-PS", "2023", "DO_NOT_SHARE_UKR_ADM2_POP_2023.csv")) %>%
```
@alaparicioc The data_dir here is defined as the `./data/` folder within the repository (i.e. published to GitHub). This file cannot be published to GitHub so needs to be stored somewhere else.

doug-leasure replied:
I made a commit to revise this.

edarin replied:
ow my data_dir is not this one. it depends on what you defined int the .env

doug-leasure replied:
`data_dir` is defined by this script to live within `repo_dir` which is the repository.  The `./data/` folder of the repo is not in the .gitgnore by design. So, those sensitive files cannot go in that location regardless of how your .env is setup.

alaparicioc replied:
The COD-PS files are in "K:\DemSci\projects\2023_WHO_Ukraine_Population\data\COD-PS", from which I pulled them.

doug-leasure committed:
modify input storage locations for 1_benchmark.R

doug-leasure commented on src/2_model/mobile_phone/1_benchmark.R:
```diff
@@ -0,0 +1,658 @@
+                    distinct(t_name, t_key, t) |>
+                    arrange(t)
+
+
```
This file (`population_baseline22.csv`) does not exist in the `./data/` directory of the repository, nor does it exist in the `in_dir` located on our internal network drive.  

@alaparicioc , we need you to provide this data so that we can run your script.

alaparicioc replied:
The dataset for COD-PS 2022 is 'ukr_admpop_2022.xlsx'. The name 'population_baseline22.csv' is an old name of the same file. I mistakenly removed the wrong lines from the script, and I did not pay too much attention since, as we agreed in our last meeting, comparing our results with COD-PS 2022 is not appropriate because COD-PS 2022 is already embedded in the methodology. I will therefore remove every reference to it.

doug-leasure reviewed (changes requested):
I made a commit with some changes and made some line-specific comments. 

I got stuck because the codps22 file in `1_benckmark.R` does not exist. Please commit that file into the `./data/` directory of the repository (this branch), and then re-request my review so that I may continue.

doug-leasure commented on src/2_model/mobile_phone/3_prepareRelease.R:
```diff
@@ -0,0 +1,33 @@
+# write output
+
+write_csv(stocks_hromada_agesex, file.path(
+  "K:/DemSci/projects/2023_WHO_Ukraine_Population/deliverables",
```
This folder is for _final_ deliverables. This output encoding also does not respect the local .env settings.  I will commit to revise to write into an `out_dir` folder.

doug-leasure committed:
deliverables to env$out_dir

alaparicioc committed:
Update 1_benchmark.R

Removing reference to COD-PS 2022 since, as we agreed in our last meeting, comparing our results with COD-PS 2022 is not appropriate because COD-PS 2022 is already embedded in the methodology.

doug-leasure committed:
tweaks to run benchmark.R

doug-leasure committed:
Merge branch 'dev' into 100-kyivstar-model-draft-rapid-response

doug-leasure committed:
Merge branch '96-vodafone-data-storing-and-formatting' into 100-kyivstar-model-draft-rapid-response

doug-leasure commented:
Note:  @edarin , I merged `dev` and `96-vodafone-data-storing-and-formatting` into this branch.

edarin committed:
merge refugee benchmark

edarin committed:
Merge branch '100-kyivstar-model-draft-rapid-response' of https://github.com/OxfordDemSci/UkraineNowPop into 100-kyivstar-model-draft-rapid-response

edarin committed:
Merge remote-tracking branch 'upstream/dev' into 100-kyivstar-model-draft-rapid-response

edarin committed:
reorder scripts

edarin committed:
adapt model to border crossing validation

edarin committed:
represent cod-ps base pyramid

edarin committed:
revisit comptying border crossing data

edarin committed:
reorder script

edarin committed:
correct returnee age-sex profile

edarin committed:
beautify missing map

edarin committed:
decompose national rescaling between outflows and internal

edarin committed:
add pcode to release

edarin committed:
cleaner pop dataset release

edarin committed:
clean plot for readme

edarin committed:
fix issue in processing of flows estimates

edarin committed:
add new batch of flows data

edarin committed:
clean script

edarin committed:
arrange estimates for new batch of data

edarin committed:
clean name of final plots

edarin committed:
add imputation step of hromada

edarin committed:
add oblast pcode to release

edarin committed:
clean missing assessment

edarin committed:
create penetrstion rate plots

edarin committed:
include subscribers in penetration rate plot

edarin committed:
prep data for dashobard with datatble

edarin committed:
penetration rate per raion

edarin committed:
rename login and update icon

edarin committed:
prep ngct vodafone data

edarin committed:
update eurostat data

edarin committed:
comment

edarin committed:
add raion info to ngct

edarin committed:
include ngct as an extra hromada

edarin committed:
make release faster

edarin committed:
run faster the penetration rate

edarin committed:
add ngct similarly as border crossing

edarin committed:
update plot for the new method

edarin committed:
remove unwanted col in release

edarin committed:
comment model

alaparicioc commented on src/2_model/mobile_phone/0_deterministic.R:
```diff
@@ -0,0 +1,1265 @@
+  filter(oblast_name_en != "Autonomous Republic of Crimea")
+
+# Prepare monthly flows
+monthlyFlows <- monthlyFlows |>
```
the flows =  1 are reported flows or you added 1 to all flows?

edarin replied:
you mean a flow of one user? I have not added anything to the reported flows.

alaparicioc commented on src/2_model/mobile_phone/0_deterministic.R:
```diff
@@ -0,0 +1,1265 @@
+
+# imputation
+# select hromadas with only three missing timesteps
+imputed_hromadas <- missing_hromadas |>
```
perhaps, it would nice to indicate the percentage of imputed hromadas?

edarin replied:
sure. I will add it to the readme.

alaparicioc commented on src/2_model/mobile_phone/0_deterministic.R:
```diff
@@ -0,0 +1,1265 @@
+## visualise interim monthly flows
+
+ggplot(
+  monthlyFlows_agesex_hat_df_stocks,
```
monthlyFlow_hat_calibrated?

alaparicioc replied:
I am unsure if this plot is important, but the Center, east, North, South and West plots look strange

<img width="1062" height="564" alt="Image" src="https://github.com/user-attachments/assets/000a139b-2f1c-4abc-bd32-5e420a4994b7" />

alaparicioc commented on src/2_model/mobile_phone/0_deterministic.R:
```diff
@@ -0,0 +1,1265 @@
+  facet_wrap(~destination_macroregion)
+
+# visualise age-sex proportion
+ggplot(
```
Why age-sex proportion 0 for 18-24 in some time steps?

<img width="1086" height="576" alt="Image" src="https://github.com/user-attachments/assets/c2034f99-e9c9-4f1f-9314-24a57e994b23" />

edarin replied:
This is the resulting age-sex proportion by macroregion and time:
<img width="743" height="621" alt="image" src="https://github.com/user-attachments/assets/5b5ae269-a5fa-4e33-ba2f-ebc65f68908c" />
None is at zero. 0.0163 is the smallest and is for male 18-24 abroad at the very beginning.

alaparicioc commented on src/2_model/mobile_phone/0_deterministic.R:
```diff
@@ -0,0 +1,1265 @@
+  labs(x = "", y = "Oblast average age-sex proportion", title = "Oblast average age-sex estimated proportion per macroregion")
+
+# Visualise age-sex profile of flows to abroad
+abroad_profile <- monthlyFlows_agesexMacroregion_hat_df |>
```
which object is this: "monthlyFlows_agesexMacroregion_hat_df"?

edarin replied:
indeed this object has disappeared. i have replaced it by the correct one.

alaparicioc reviewed (commented):
I have left some comments on the model.

alaparicioc commented on src/2_model/mobile_phone/1_benchmark.R:
```diff
@@ -0,0 +1,880 @@
+idp <- data_raw$result
+
+
+# Compare numbers
```
I do not remember to add this part onwards. So, I did not make any change from now onwards

alaparicioc reviewed (commented):
I added more comments

alaparicioc committed:
Update 1_benchmark.R

updated code for new estimates

alaparicioc commented on src/2_model/mobile_phone/2_prepareRelease.R:
```diff
@@ -0,0 +1,100 @@
+)
+
+# load data
+pcodes <- fread(file.path(here::here("src/dashboard/api/app/data/db-data/global_pcodes.csv")))
```
it works for me when data.table::fread

edarin replied:
i dont understand - data.table library should be loaded already as it is the first line of the code.

alaparicioc commented on src/2_model/mobile_phone/2_prepareRelease.R:
```diff
@@ -0,0 +1,100 @@
+
+# prepare flows data
+flows[, origin_hromada_PCODE :=
+  fifelse(origin_hromada == "Kyiv", "UA8000000", origin_hromada)]
```
ifelse?

edarin replied:
fifelse is the data.table version of ifelse

alaparicioc commented on src/2_model/mobile_phone/2_prepareRelease.R:
```diff
@@ -0,0 +1,100 @@
+]
+
+flows[, origin_raion_PCODE :=
+  fifelse(origin_raion == "Kyiv", "UA8000", origin_raion)]
```
ifelse?

edarin replied:
fifelse is the data.table version of ifelse

alaparicioc commented on src/2_model/mobile_phone/1_prepareRelease.R:
```diff
@@ -0,0 +1,100 @@
+  destination_raion_PCODE := str_sub(destination_raion_PCODE, 1, 6)
+]
+
+setnames(flows, "monthlyFlow_hat_calibrated", "pop_estimated")
```
is there any library for this function?

edarin replied:
which function? setnames? this is from the data.table package loaded at the beginning of the script.

alaparicioc commented on src/2_model/mobile_phone/2_prepareRelease.R:
```diff
@@ -0,0 +1,100 @@
+    raion = destination_raion,
+    hromada_PCODE = destination_hromada_PCODE,
+    raion_PCODE = destination_raion_PCODE,
+    oblast_PCODE = destination_oblast_PCODE
```
i did not get the oblast column

edarin replied:
I ran again and had the oblast column

alaparicioc reviewed (commented):
added more comments

doug-leasure reviewed (approved):
I'm not going to be able to review all of the code. I will be happy once @alaparicioc and @edarin have agreed and finalised the code. I will prioritise reviewing/editing the data and release statement.

doug-leasure commented:
I put all of the release materials together in a folder on Google Drive: 
https://drive.google.com/drive/folders/1_z9yNw9NG962Fhf1bYi3fRbyDzLll0p1?usp=drive_link

edarin committed:
standardise output and change design

edarin committed:
plot unique combinations

edarin committed:
write output per month

edarin committed:
enhance plots

edarin committed:
correct dip in raw data

edarin committed:
adapt idp comparison

edarin committed:
plot origin of flows

edarin committed:
make release more compliant with cod-ps

doug-leasure reviewed (changes requested):
Thanks for updating the data format for the release. This is much better! 

A few minor changes please for the data:
- Rename column "t" to "DATE"
- Rename age-sex columns to match CODPS format (e.g. "F_18_24"; note the underscore rather than hyphen)
- Rename the columns reflecting all ages to remove "TL" wihch should only be used to refer to total population. Please use "F_18_64", for example.

edarin committed:
impute vodafone before ratio

edarin committed:
change order - change api dtm

edarin committed:
add metrics to output

edarin committed:
move imputation to data stage

edarin committed:
remove data for 2025-06-01

edarin committed:
change t to DATE

edarin committed:
make internal origin plot faster

edarin committed:
implement ratio on stayers

edarin committed:
change ref date to avoid august

edarin committed:
change the place of agg

edarin committed:
add options of sample

edarin committed:
update last date
